packages/zoltan2/core/src/algorithms/color/Zoltan2_AlgHybridD1.hpp

```diff
@@ -0,0 +1,985 @@
+	Kokkos::deep_copy(verts_to_send_size_host, verts_to_send_size);
+        gno_t recv, sent;
+        comm_time = doOwnedToGhosts(mapOwnedPlusGhosts,
+			            nVtx,
+				    verts_to_send_host,
+				    verts_to_send_size_host,
+				    femv,
+				    recv,
+				    sent);
+        sentPerRound[0] = sent;
+        recvPerRound[0] = recv;
+        if(verbose) std::cout<<comm->getRank()<<": done communicating\n";
+	verts_to_send_size_host(0) = 0;
+        deep_copy(verts_to_send_size, verts_to_send_size_host);
+        //set the old ghost colors
+        Kokkos::parallel_for(rand.size()-nVtx,KOKKOS_LAMBDA(const int& i){
+          ghost_colors(i) = femv_colors(i+nVtx);
+        });
+	Kokkos::fence();
+	//detect conflicts on the device, uncolor conflicts consistently.
+        double temp = timer();
+	detectConflicts<execution_space, memory_space>(nVtx,
+			                               rand.size()-nVtx,
+						       dist_offsets,
+						       dist_adjs,
+						       femv_colors,
+						       verts_to_send_atomic,
+						       verts_to_send_size_atomic,
+						       recoloringSize,
+						       rand_dev,
+						       gid_dev,
+						       ghost_degrees_dev,
+						       recolor_degrees);
+        deep_copy(recoloringSize_host, recoloringSize);
+        conflict_detection += timer() - temp;
+        comp_time += conflict_detection;
+      }
+      //end the initial recoloring
+      if(verbose)std::cout<<comm->getRank()<<": done initial recoloring, begin recoloring loop\n";
+      double totalPerRound[100];
+      double commPerRound[100];
+      double compPerRound[100];
+      double recoloringPerRound[100];
+      double conflictDetectionPerRound[100];
+      double serialRecoloringPerRound[100];
+      int vertsPerRound[100];
+      bool done = false; //We're only done when all processors are done
+      if(comm->getSize() == 1) done = true;
+      totalPerRound[0] = interior_time + comm_time + conflict_detection;
+      recoloringPerRound[0] = 0;
+      commPerRound[0] = comm_time;
+      compPerRound[0] = interior_time + conflict_detection;
+      conflictDetectionPerRound[0] = conflict_detection;
+      recoloringPerRound[0] = 0;
+      vertsPerRound[0] = 0;
+      int distributedRounds = 1; //this is the same across all processors
+      int serial_threshold = this->pl->template get<int>("serial_threshold",0);
+      
+      Kokkos::View<lno_t*, device_type> verts_to_recolor("verts_to_recolor", boundary_size);
+      typename Kokkos::View<int*, device_type>::HostMirror ghost_colors_host;
+      //while the queue is not empty
+      while(recoloringSize_host(0) > 0 || !done){
+	if(recoloringSize_host(0) < serial_threshold) break;
+        //get a subview of the colors:
+        auto femvColors = femv->getLocalViewDevice();
+        auto femv_colors = subview(femvColors, Kokkos::ALL, 0);
+        //color everything in the recoloring queue, put everything on conflict queue
+        if(distributedRounds < 100) {
```

Comment:
This looks s bit dangerous. Perhaps you need to check the length of vertsPerRound, or is that always > 100? If so, better to define a const variable for this.

Reply:
These statistic arrays are all statically allocated to be size 100, so this is technically safe. I will add a const variable as you suggested, because magic numbers are dangerous, and harmful to readability.